win32ss/user/ntuser/ghost.c

```diff
@@ -128,6 +216,119 @@ HWND FASTCALL UserHungWindowFromGhostWindow(HWND hwndGhost)
     return IntHungWindowFromGhostWindow(pGhostWnd);
 }
 
+BOOL IntHaveToQuitGhosting(void)
+{
+    RTL_ATOM ClassAtom;
+    HWND hwndDesktop;
+    PWND pDesktopWnd;
+    UNICODE_STRING WindowName;
+
+    if (!IntGetAtomFromStringOrAtom(&GhostClass, &ClassAtom))
+        return FALSE;
+
+    hwndDesktop = IntGetCurrentThreadDesktopWindow();
+    pDesktopWnd = ValidateHwndNoErr(hwndDesktop);
+    if (!pDesktopWnd)
+        return FALSE;
+
+    RtlInitUnicodeString(&WindowName, NULL);
+
+    return IntFindWindow(pDesktopWnd, NULL, ClassAtom, &WindowName) != NULL;
+}
+
+VOID NTAPI
+GhostThreadProc(_In_ PVOID StartContext)
+{
+    HWND hwndTarget = (HWND)StartContext;
+    MSG msg;
+
+    if (!IntCreateGhostWindow(hwndTarget))
+    {
+        DPRINT1("Unable to create ghost\n");
+        goto Quit;
+    }
+
+    // thread message loop
+    for (;;)
+    {
+        RtlZeroMemory(&msg, sizeof(MSG));
+
+        if (!co_IntGetPeekMessage(&msg, NULL, 0, 0, PM_REMOVE, TRUE))
+            break;
+
+        switch (msg.message)
+        {
+            case PM_CREATE_GHOST:
+            {
+                hwndTarget = (HWND)msg.wParam;
+                if (!IntCreateGhostWindow(hwndTarget))
+                {
+                    DPRINT1("Unable to create ghost\n");
+                }
+                break;
+            }
+        }
+
+        IntTranslateKbdMessage(&msg, 0);
+        IntDispatchMessage(&msg);
+
+        if (IntHaveToQuitGhosting())
+        {
+            DPRINT1("Have to quit ghost thread\n");
+            break;
+        }
+    }
+
+Quit:
+    RtlFreeHeap(GlobalUserHeap, 0, gGhostInfo);
+
+    KeSetEvent(&gGhostInfo->GhostQuitEvent, IO_NO_INCREMENT, TRUE);
+
+    gGhostInfo = NULL;
+}
+
+BOOL IntAddGhost(HWND hwndTarget)
+{
+    NTSTATUS Status;
+    CLIENT_ID ClientId;
+    HANDLE GhostThreadHandle;
+
+    if (gGhostInfo)
+    {
+        return NtUserPostThreadMessage(gGhostInfo->ThreadId, PM_CREATE_GHOST,
+                                       (WPARAM)hwndTarget, 0);
+    }
+
+    gGhostInfo = RtlAllocateHeap(GlobalUserHeap, HEAP_ZERO_MEMORY, sizeof(GHOST_INFO));
+    if (!gGhostInfo)
+        return FALSE;
+
+    gGhostInfo->hwndTarget = hwndTarget;
+
+    // create events
+    KeInitializeEvent(&gGhostInfo->GhostStartupEvent, SynchronizationEvent, FALSE);
+    KeInitializeEvent(&gGhostInfo->GhostQuitEvent, SynchronizationEvent, FALSE);
+
+    // create thread
+    Status = PsCreateSystemThread(&GhostThreadHandle,
+                                  STANDARD_RIGHTS_ALL,
+                                  ___ObjectAttributes___,
```

Comment:
What's this? Does it compile?

Reply:
Doesn't seem so according to the buildbots:
```
../../win32ss/user/ntuser/ghost.c: In function 'IntAddGhost':
../../win32ss/user/ntuser/ghost.c:315:35: error: '___ObjectAttributes___' undeclared (first use in this function)
```

Reply:
You want to use either InitializeObjectAttributes or RTL_CONSTANT_OBJECT_ATTRIBUTES. Your thread doesn't need a name, and the only flag needed should be the kernel handle one, so this should do:
`OBJECT_ATTRIBUTES ObjectAttributes = RTL_CONSTANT_OBJECT_ATTRIBUTES(NULL, OBJ_KERNEL_HANDLE);`